src/KubernetesClient/Informers/ResourceObservableExtensions.cs

```diff
@@ -0,0 +1,294 @@
+        /// </summary>
+        /// <param name="source">The source sequence</param>
+        /// <param name="cache">The cache to synchronize</param>
+        /// <param name="keySelector">The key selector function</param>
+        /// <typeparam name="TKey">The type of key</typeparam>
+        /// <typeparam name="TResource">The type of resource</typeparam>
+        /// <returns>Source sequence wrapped into <see cref="CacheSynchronized{T}" />, which allows downstream consumers to synchronize themselves with cache version</returns>
+        public static IObservable<CacheSynchronized<ResourceEvent<TResource>>> SynchronizeCache<TKey, TResource>(
+            this IObservable<ResourceEvent<TResource>> source,
+            ICache<TKey, TResource> cache,
+            Func<TResource, TKey> keySelector,
+            Func<TResource, TResource, TResource> mapper = null)
+        {
+            if (mapper == null)
+                mapper = (oldResource, newResource) => newResource;
+            long msgNum = 0;
+            return source
+                .Do(_ => msgNum++)
+                .WithResets(events =>
+                {
+                    var reset = events
+                        .Select(x => x.Value)
+                        .Where(x => x != null)
+                        .ToDictionary(keySelector, x => x);
+
+                    cache.Reset(reset);
+                    cache.Version += events.Count;
+                    return events;
+                }, notification =>
+                {
+                    if (notification.EventFlags.HasFlag(EventTypeFlags.Reset))
+                    {
+                    }
+                    else if (!notification.EventFlags.HasFlag(EventTypeFlags.Delete))
+                    {
+                        cache.Version++;
+                        if (notification.Value != null)
+                        {
+                            var key = keySelector(notification.Value);
+                            if (cache.TryGetValue(key, out var existing))
+                                notification = new ResourceEvent<TResource>(notification.EventFlags, mapper(existing, notification.Value), existing);
+                            cache[keySelector(notification.Value)] = notification.Value;
+                        }
+                    }
+                    else
+                    {
+                        cache.Remove(keySelector(notification.OldValue));
+                    }
+
+                    return new CacheSynchronized<ResourceEvent<TResource>>(msgNum, cache.Version, notification);
+                });
+        }
+
+
+        public static IObservable<ResourceEvent<TResource>> ComputeMissedEventsBetweenResets<TKey, TResource>(this IObservable<ResourceEvent<TResource>> source, Func<TResource, TKey> keySelector, IEqualityComparer<TResource> comparer)
+        {
+            Dictionary<TKey, TResource> cacheSnapshot;
```

Comment:
These variables will be shared across all subscriptions and I do not believe that is correct.

Their declarations should be moved into the lambda supplied to Create.